### PR TITLE
search frontend: refine revision hovers

### PR DESCRIPTION
**Pull request by rvantonder**
Hovers shouldn't be misleading, so I wanted to support the full spectrum of revision syntax. I made some ugly discoveries, documented in #16628. This PR handles all the tricky/obscure details of the revision syntax in hovers. In time it'd be great to change some of the syntax conventions here.

Example: `*` or `*!` at the beginning of a revision string now conveys the true meaning. This is likely not a heavily used syntax/feature, but we do need to be principled and explicit about what's going on here, so that if/when we change things we will have a place in the code logic + tests that tracks it.

Screen Shot 2020-12-10 at 1 51 23 PM

Part of  #16141.

### CODE REVIEW

File: client/shared/src/search/query/hover.ts [hunk 1]
```diff
@@ -110,26 +111,31 @@ const toRevisionHover = (token: MetaRevision): string => {
...
+        case MetaSourcegraph.ExcludeGlobMarker:
+            title = 'glob pattern to exclude'
+            description =
+                'A prefix indicating that git references, like a commit or branch name, should be **excluded** from search based on the glob pattern that follows. Used in conjunction with a glob pattern that matches a set of commits or branches, followed by a a pattern to exclude from the set. For example, `*refs/heads/*:*!refs/heads/release*` searches all branches at the head commit, excluding branches matching `release*`.'
+            break
+        case MetaSourcegraph.Separator:
+            title = 'separator'
+            description =
+                'Separates multiple revisions to search across. For example, `1a35d48:feature:3.15` searches the repository for matches at commit `1a35d48`, or a branch named `feature`, or a tag `3.15`.'
+            break
```
Comment (lguychard) [inline client/shared/src/search/query/hover.ts:138]: Minor (and maybe overkill if it's not destined to grow any further), but I'm not a huge fan of this evergrowing `switch` statement, mixing data with imperative logic.

Maybe the titles/descriptions could be extracted into a `Record<MetaGit | MetaSourcegraph, { title: string, description: string}>`?

```typescript
const revisionHoverContents: Record<MetaGit | MetaSourcegraph, { title: string, description: string }> = {
    [MetaSourcegraph.Separator]: {
        title: 'separator',
        description: 'Separates multiple revisions to search across. For example, `1a35d48:feature:3.15` searches the repository for matches at commit `1a35d48`, or a branch named `feature`, or a tag `3.15`.'
    },
    ...
}

const toRevisionHover = (token: MetaRevision): string => {
    if (token.kind === MetaGit.Label && token.value.match(/^head$/i)) {
        return '**Revision HEAD**. Search the repository at the latest HEAD commit of the default branch.'
    }
    const { title, description } = revisionHoverContents[token.kind]
    return `**Revision ${title}**. ${description}`
}
```

File: client/shared/src/search/query/decoratedToken.ts [hunk 2]
```diff
@@ -81,21 +81,34 @@ export interface MetaField extends BaseMetaToken {
...
+export type MetaRevisionKind = MetaGit | MetaSourcegraph
+
+/**
+ * A custom revision syntax that is only valid in Sourcegraph search queries.
+ */
+export enum MetaSourcegraph {
+    Separator = 'Separator', // a ':' that separates revision patterns
+    IncludeGlobMarker = 'IncludeGlobMarker', // a '*' at the beginning of a revision pattern to mark it as an "include" glob pattern.
+    ExcludeGlobMarker = 'ExcludeGlobMarker', // a '*!' at the beginning of a revision pattern to mark it as an "exclude" glob pattern.
+}
```
Comment (lguychard) [inline client/shared/src/search/query/decoratedToken.ts:102]:
```suggestion
export enum MetaSourcegraphRevision {
    Separator = 'Separator', // a ':' that separates revision patterns
    IncludeGlobMarker = 'IncludeGlobMarker', // a '*' at the beginning of a revision pattern to mark it as an "include" glob pattern.
    ExcludeGlobMarker = 'ExcludeGlobMarker', // a '*!' at the beginning of a revision pattern to mark it as an "exclude" glob pattern.
}
```

Nit: a more specific name would make this explicit that this enum solely refers to revision-specific syntax (and not **all** Sourcegraph syntax)

**Review by lguychard — APPROVED** [review]

File: client/shared/src/search/query/decoratedToken.ts (hunk 2 above)
Comment (rvantonder) [reply client/shared/src/search/query/decoratedToken.ts:102]: cool, added it to get `MetaGitRevision` too.

File: client/shared/src/search/query/hover.ts (hunk 1 above)
Comment (rvantonder) [reply client/shared/src/search/query/hover.ts:138]: > mixing data with imperative logic. 

 problem? 

but seriously, yeah. So actually what I think is, just going to put `return` for each case branch. So what if there's a bit of redundant `**Revision **.`, it's actually not a big deal?

**Pull request merged**